File: In Class Assignments/Ex5/ex5Class.R

################################################################################

setwd("In Class Assignments\\Ex5")
require(TMB)

################################################################################

m <- scan("ex5.dat",skip=1,n=1,quiet=T)
TT <- scan("ex5.dat",skip=3,n=m,quiet=T)
Tmax <- scan("ex5.dat",skip=5,n=1,quiet=T)
B <- matrix(scan("ex5.dat",skip=7,n=m*Tmax,quiet=T),ncol=Tmax,byrow=T)
R <- matrix(scan("ex5.dat",skip=58,n=m*Tmax,quiet=T),ncol=Tmax,byrow=T)
Phi0 <- scan("ex5.dat",skip=109,n=m,quiet=T)

data <- list(m=m,TT=TT,Tmax=Tmax,B=B,R=R,Phi0=Phi0)

################################################################################
# Hint Nproj = 0 for basic estimation
# Basic estimation
################################################################################

# Set the parameters to initial values
#bounds are on ex 5 document
# parameters <- list(dummy=0,mu=-1,log_tau=2,B0=rep(500, 50),
#                    log_sigR=rep(log(2),50),eta=rep(0.2,50))


parameters <- list(dummy=0,mu=0,log_tau=-1.5,B0=rep(1000.1/2, 50),
                   log_sigR=rep(-1.5,50),eta=rep(0,50))


#print(data)
#print(parameters)

compile("Ex5Class.cpp", flags="-Wno-ignored-attributes")
dyn.load(dynlib("Ex5Class"))

# test code - for checking for minimization
map<-list(mu=factor(NA),log_tau=factor(NA),B0=rep(factor(NA),m),log_sigR=rep(factor(NA),m),eta=rep(factor(NA),m))
model <- MakeADFun(data, parameters, DLL="Ex5Class",silent=T,map=map)
xx <- model$fn(model$env$last.par)
cat(xx,model$report()$obj_fun,"\n")
#AAA

# Now estimate everything
map<-list(dummy=factor(NA))
model <- MakeADFun(data, parameters, random="eta", DLL="Ex5Class",silent=T,map=map)

rep <- sdreport(model)
rep
# Bounds on the parameters
# parameters <- list(dummy=0,mu=-1,log_tau=2,B0=rep(500, 50),
#                    log_sigR=rep(log(2),50),eta=rep(5,50))

lowbnd= c(mu= -10, log_tau = -7, B0= rep(0.1,50),
         log_sigR = rep(-7, 50), eta= rep(-Inf, 50))

uppbnd= c(mu =  10, log_tau = 4, B0 = rep(1000,50),
          log_sigR = rep(4, 50), eta = rep(Inf, 50))

fit <- nlminb(model$par, model$fn, model$gr,
              control=list(rel.tol=1e-12,eval.max=100000,iter.max=1000),
              lower=lowbnd,upper=uppbnd)


best <- model$env$last.par.best
print(best)
rep <- sdreport(model)
print(summary(rep))

fit$objective

File: In Class Assignments/Ex5/Ex5Class.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(m)
  DATA_IVECTOR(TT)
  DATA_INTEGER(Tmax)
  DATA_MATRIX(B)
  DATA_MATRIX(R)
  DATA_VECTOR(Phi0)

  // End of data section

  PARAMETER(dummy);
  PARAMETER(mu);
  PARAMETER(log_tau);
  PARAMETER_VECTOR(B0);
  PARAMETER_VECTOR(log_sigR);
  PARAMETER_VECTOR(eta);
  // End of the estimated parameters

  vector<Type> R0(m);
  vector<Type> SigR(m);
  vector<Type> h(m);
  Type tau;
  Type beta;
  Type BH;
  Type obj_fun;
  matrix<Type> BHout(Tmax,m);
  // End of the temporary variables

  // Transform the parameters
  R0 = B0/Phi0;
  tau = exp(log_tau);
  SigR = exp(log_sigR);

  obj_fun = 0;
 for (int k=0;k<m;k++) {

   // Extract beta and define h
   beta = mu + tau*eta(k);
   
   h(k) = (exp(beta + 0.2))/ (exp(beta + 1.0));
  
   for(int t = 0; t < Tmax; t++){
     BH = (4.0 * R0(k) * h(k) * B(k,t))/((Phi0(k)*R0(k)*(1.0-h(k)))+ (5.0*h(k)-1.0)*B(k,t));
      
     BHout(t,k) = BH;
   // Likelihood
   
   obj_fun += square(log(R(k,t))-log(BH) + square(SigR(k))/2.0)/(square(SigR(k))) + log(SigR(k));
   
   }
   
   obj_fun -= dnorm(eta(k), Type(0), Type(1), true);

   
  }



  obj_fun += dummy*dummy;
  ADREPORT(h);
  ADREPORT(R0);
  ADREPORT(tau);
  REPORT(BHout);

  return(obj_fun);
}
